### PR TITLE
[3.4] backport for fix retry requests when receiving ErrGPRCNotSupportedForLearner

### PR DESCRIPTION
**Pull request by sheyt0**
Backport of PR https://github.com/etcd-io/etcd/pull/12985

### CODE REVIEW

**Comment by ivanvc** [comment]
The `release-3.4` branch doesn't work yet with `/ok-to-test`. Once #17619 gets merged, we can bring it to 3.4. Can a maintainer manually approve the workflow? cc. @jmhbnz, @ahrtr

**Review by ahrtr — APPROVED** [review]

**Review by jmhbnz — APPROVED** [review]

**Pull request merged**
